File: 36-valid-sudoku/36-valid-sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        def validate(s):
            res = [0] * 9
            for char in s:
                if '1'<= char <= '9':
                    res[int(char)-1] += 1
            if max(res) != 0 and max(res) > 1:
                return False
            return True
        for row in board:
            if not validate(row):
                return False
        for col in range(9):
            if not validate([board[l][col] for l in range(9)]):
                return False
        for k in range(9):
            stack = []
            for i in range(k%3*3,3*(k%3+1)):
                for j in range(k//3*3,k//3*3+3):
                    stack.append(board[i][j])
            if not validate(stack):
                return False
        return True
                
            
        

File: 36-valid-sudoku/36-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        //row examination
        boolean isRowValid = isRowColValid(board, true);
        //col examination
        boolean isColValid = isRowColValid(board, false);
        //sub-boxes examination
        boolean isBoxValid = isSubValid(board);
        return isRowValid && isColValid && isBoxValid;
    }
    
    public boolean isRowColValid(char[][] board, boolean isRow){
        int nrow = board.length, ncol = board[0].length;
        boolean isRowValid = true;
        Set<Character> existed = null;
        if(!isRow){
            int temp = nrow;
            nrow = ncol;
            ncol = temp;
        }
        Character val = null;
        for(int row = 0; row < nrow; row++){
            existed = new HashSet<>();
            for(int col = 0; col < ncol; col++){
                if(isRow){
                    val = board[row][col];
                }else{
                    val = board[col][row];
                }
                if(!(Character.isDigit(val)))continue;
                if(existed.contains(val)){
                    isRowValid = false;
                    return isRowValid;
                }
                existed.add(val);
            }
            
        }
        return isRowValid;  
    }
    public boolean isSubValid(char[][] board){
        int nrow = board.length, ncol = board[0].length;
        Set<Character> existed = null;
        boolean ans = true;
        for(int col = 0; col < ncol; col += 3)
            for(int row = 0; row < nrow; row += 3){
                existed = new HashSet<>();
                for(int i = row; i < row + 3; i++)
                    for(int j = col; j < col + 3; j++){
                        if(!(Character.isDigit(board[i][j])))continue;
                        if(existed.contains(board[i][j]))return false;
                        existed.add(board[i][j]);
                    }
            }
        
        return true;
        
    }
    
}